\documentclass{article}

\usepackage{caption,color,fancyvrb,subfig}
%\usepackage{code,algorithmic,algorithm}
%\usepackage{algpseudocode}
\usepackage{algorithm}
\usepackage[noend]{algpseudocode}
%\usepackage[]{algorithm2e}
%\usepackage{graphicx,epsfig}
\usepackage{amsmath,amsthm,amsfonts}
%\usepackage{url}
\usepackage{graphicx}
\usepackage{setspace}
\usepackage{verbatim}
\usepackage{datetime}
\usepackage{float}
\usepackage{booktabs} 
\usepackage{tablefootnote}
\usepackage{multirow} 
\usepackage{amsthm}
\usepackage{float}
\newfloat{algorithm}{t}{lop}
 \floatname{algorithm}{Algorithm}
\newtheorem{theorem}{Theorem}[section]
\newtheorem{lemma}[theorem]{Lemma}
\newtheorem{definition}[theorem]{Definition}

\usepackage[backend=bibtex8,style=numeric,url=false,isbn=false]{biblatex} 

\renewbibmacro{in:}{}

\addbibresource{clustering_tools.bib}
\addbibresource{metagenomics.bib}
\addbibresource{string.bib}
\addbibresource{extra.bib}


%\begin{comment}
\textwidth = 6.5 in
\textheight = 9 in
\oddsidemargin = 0.0 in
\evensidemargin = 0.0 in
\topmargin = 0.0 in
\headheight = 0.0 in
\headsep = 0.0 in
\parskip = 0.2in
\parindent = 0.0in
%\end{comment}

\def\inv{^{-1}}

\newcommand{\ignore}[1]{}
\newdateformat{mydate}{\monthname[\THEMONTH] \ordinaldate{\THEDAY}, \THEYEAR}

\begin{document} 




\begin{titlepage}
\begin{center}

\textsc{\huge \bfseries \sc{Algorithms for clustering highly conserved phylogenetic markers}}\\
%\textsc{\huge \bfseries \sc{Title Line 2}}\\[1.0cm]
 
\emph{A prospectus submitted in partial fulfillment of the degree of Doctor of
Philosophy}\\[5.5cm]

%\textsc{\large DRAFT}\\
%\emph{October 18, 2011}\\[2.0cm]

\emph{Preliminary Oral Examination for}\\
\textsc{\large Christopher Michael Hill}\\[2.0cm] % [4.0cm]
\emph{Advisor:} \\
\textsc{Dr. Mihai Pop}\\[.5cm]
\emph{Committee Members:}\\
\textsc{Dr. Atif Memon}\\
\textsc{Dr. H\'ector Corrada Bravo}\\[4.0cm]

{\bfseries Department of Computer Science}\\
{\bfseries University of Maryland, College Park, MD 20742}\\
{\bfseries \mydate\today}
\vfill

\end{center}
\end{titlepage}

\newpage
\thispagestyle{empty}
\mbox{}

\setcounter{page}{1}
\doublespacing
\begin{abstract}

Microbes play a large role in every aspect of our lives.
They aid in digestion of food, have been associated with different diseases, and even help identify which keyboard and mouse we use.
In the environment, these microbes are responsible for most of the biochemical cycles that allow life to exist on earth, such as the carbon and nitrogen cycle.
Despite their importance, these organisms are among the least understood on Earth due in part to the difficult nature of culturing them in a laboratory setting.
Fortunately, advances in sequencing technology allow us to capture a small, noisy snapshot of the microbial community.
In typical analyses, the short fragments produced from these sequencing technologies first need to be clustered into similar groups based on some definition of similarity.
%The vast amount of sequencing data has led to researchers utilizing heuristics to cluster this data.

In this prospectus, we will describe the current approaches and limitations for clustering biological sequences, which are represented electronically as strings of characters.
First, we will discuss multi-core algorithms that speedup the runtime of clustering.
Then, we will describe efficient string matching algorithms to quickly recruit sequences to potential centers.
By sorting highly similar sequences, we provide a novel way to reduce the amount of work done during cluster recruitment.
Lastly, we will discuss the problematic issue of sequences that can align equally well to multiple centers and present strategies for handling them.


\end{abstract}
\setstretch{.5}

\newpage
\tableofcontents

\newpage
\listoffigures
\listoftables
\listofalgorithms

\newpage
\doublespacing


%\section{Introduction}
\input{introduction.tex}

%\begin{figure}
%\begin{center}
%\includegraphics[scale=1.0]{img/sample_figure}
%\end{center}
%\caption{Sample camption}
%\label{fig:sample_figure}
%\end{figure}

%\begin{figure}
%\begin{center}
%\begin{verbatim}
%/* Sample code snippet */
%\end{verbatim}
%\end{center}
%\caption{Code snippet caption}
%\label{fig:sample_code}
%\end{figure}


%\section{Related Work}
\input{related_works.tex}

%\section{Preliminary Work}
\input{preliminary_work.tex}

%\section{Proposed Work}
%\input{proposed_work.tex}

\section{Timeline}

\begin{center}
\begin{tabular}{|l|c|}
\hline
Parallel DNACLUST & 1 months \\
Fast, exact clustering tool using efficient $k$-diff algorithm & 2-4 months \\
Handling ambiguous read and their affect on differential abundance & 1 months \\
\hline
\textbf{TOTAL} & 4-6 months \\
\hline
\end{tabular}
\end{center}

Paper deadline goals:

\begin{itemize}
\item 14th Workshop on Algorithms in Bioinformatics (WABI), May 2014: Parallel DNACLUST and ambiguous read assignment strategy
\item Nucleic Acids Research, September 2014: Efficient k-difference clustering tool
\end{itemize}


\section{Conclusion}

Microorganisms play an important role in the environment.
Despite this, these microbes are among the least understood on Earth due in part to difficulties culturing them in a lab setting.
Sequencing technology provides us with a solution, allowing us to get a snapshot of the microscopic communities that inhabit these environments.
Unfortunately, this snapshot only can provide us with a very small, noisy picture of the community.
In typical analyses, the short fragments of DNA produced from sequencing technologies need to be clustered into similar groups based on some similarity.
In this prospectus, we have described the current approaches and limitations for clustering highly similar DNA sequences.
By leveraging efficient string matching algorithms and multi-core systems, we have proposed a fast and exact clustering tool for DNA sequences.

\newpage


\section{Acknowledgments}

I owe my sincere gratitude to all the people who made this work possible.
I'd like to thank my adviser Mihai Pop for giving me the opportunity to work on intellectually challenging and interesting projects for nearly 8 years.
His guidance and support has led me to pursue a career in academia and I am forever grateful.
Furthermore, I would like to thank my committee members and mentors Atif Memon and H\'ector Corrada Bravo for their support.
I owe a great thanks to all the members of Pop, Bravo, and Memon labs for their discussions, contribution, and friendship.

\appendix
\section{Reading List}
\label{app:readinglist}

\subsection{Clustering}

%\addbibresource{clustering_tools.bib}
\printbibliography[keyword=chriscluster,heading=subbibliography,heading=none]

\subsection{String algorithms}

\printbibliography[keyword=chrisstring,heading=subbibliography,heading=none]

\subsection{Metagenomics}

\printbibliography[keyword=chrismetagenomics,heading=subbibliography,heading=none]


\newpage

\nocite{*}
%\printbibliography

%\bibliographystyle{plain} 
%\bibliography{prelim}


\end{document}

